### ISSUE TITLE
Whether horovod implements cross iteration scheduling

### ISSUE THREAD

**Issue by Tron-x** [horovod/horovod#1558]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version:
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Your question:**
Hi, Whether horovod implements cross iteration scheduling？
Or can we flexibly implement layer-wise communication and computation overlap in this layer?

**Comment by tgaddair** [comment]
Hey @Tron-x, Horovod does not currently support cross-iteration allreduce, but contributions are welcome if you have in mind a particular approach.  At NeurIPS this year, there was an interesting poster that @abditag2 showed me about applying allreduced gradients from the previous step to update the model in the current step, allowing the allreduce to occur in the background, which sounded interesting.

Layer-wise communication and computation overlap could be interesting, though as I understand it would be difficult to implement in some frameworks due to constraints like TensorFlow's global barrier.  I understand that some people have had success doing this, but I haven't seen their implementations personally, and am not sure how generalizable their solutions were.

**Comment by Tron-x** [comment]
Hey@Travis Addair ，I'm interested in layer wise computation and communication overlap, but it seems to be difficult to implement. I see that bytescheduler implements PS cross iteration https://github.com/bytedance/byteps/blob/bytescheduler/bytescheduler/bytescheduler/tensorflow/plugin.cc. I'm confused about whether you think cross iteration will bring performance improvement? If I plan to implement cross iteration allreduce, can you give me some suggestions on code implementation？ Or which part of the current horovod improvement will bring performance improvement?

**Comment by tgaddair** [comment]
Hey @Tron-x, I suspect that it could give performance improvements, but only in certain situations.  For instance, in cases where your network cannot keep up with your GPUs, but your model parameters are small enough that you can keep some tensors on the device for allreduce while a forward pass is happening at the same time.

In the BytePS approach, I believe they get around this by having all the gradient aggregation occur in host memory, not GPU memory.  So there's no contention, but as a tradeoff they need to copy memory out of the GPU, which can slow things down.

We can of course do the same by using MPI instead of NCCL, but in practice when you have good network, NCCL significantly outperforms MPI.

I think it would be worth experimenting with for models that don't scale well using a traditional allreduce approach, like VGG16.  Would like to get @romerojosh's thoughts on this as well.

As for how you could implement this in Horovod, most of the code would probably need to live in the [Controller](https://github.com/horovod/horovod/blob/master/horovod/common/controller.cc), which decides what tensors are going to be allreduced each time.  You may also need to write some logic in the [TensorFlow custom op](https://github.com/horovod/horovod/blob/master/horovod/tensorflow/mpi_ops.cc) for TensorFlow specific integrations.

**Comment by romerojosh** [comment]
I may be misunderstanding the question here but fundamentally, there is nothing in the current Horovod backend that would disallow cross iteration scheduling of collective operations. The reason for this is that the backend has no concept of an iteration: it simply checks at a given regular interval what tensor collective operations are submitted by all the workers, does an intersection of these tensors, and proceeds to carry out operations that are common across all workers. In other words, the backend does not know, or care, if the operations queued are from the current step, the next step, or something else altogether. So long as the tensors are being submitted on all workers, it will perform the common collective operations.

In addition to this, Horovod enforces no explicit global barriers (on GPU or CPU) that would block the framework (all barriers either impact only the Horovod background thread or the independent Horovod CUDA stream), and only interacts with the framework through callback functions to indicate pending work has completed. Thus, any blocking is driven solely through the frameworks enforced dependencies and scheduler. 

As a result, if the framework itself is capable cross iteration scheduling, Horovod _should_ handle this already. However, if the framework enforces its own global barriers between steps (on the CPU or GPU), then Horovod cannot do anything about that.

**Comment by Young768** [comment]
@tgaddair Hi, I found a paper related to this idea. The basic idea is to synchronize partial results among workers. For example, we have two workers, fast w0 and slow w1. When we call the allreduce operation, only the gradients from worker w0 is ready, i.e. g0. Since w1 has no output yet, it will synchronize a NULL with g0 via the allreduce. In the next iteration, if it can catch up with other processes, then they can join the synchronization. I'd like to implement this idea and contribute code to Horovod. Is the related code in this file, [Controller](https://github.com/horovod/horovod/blob/master/horovod/common/controller.cc),?

**Comment by Young768** [comment]
Hi, @tgaddair "Once a tensor is ready to be reduced, the coordinator sends a Response instructing all ranks to start the reduction to all ranks." Does the function ConstructResponse(std::string& name, int joined_size = 0); check if tensors from all the rank are ready and then start the reduction?

**Comment by Tron-x** [comment]
Can we talk about it? I'm interested in it，can you give me your email?

**Comment by tgaddair** [comment]
Hey @Young768, thanks for looking into this!  I think your approach makes sense.  For frameworks like TensorFlow and PyTorch, one way to accomplish what you suggest would be to have the `hvd.allreduce` op return immediately with a zero tensor (no gradient update) the first time it's called, then always return the result of the previous step's allreduce afterwards.

When the allreduce completes, you would want to place it on some kind of results queue.  This can be done as part of the callback passed to `EnqueueTensorAllreduce`.  Then, the next time the allreduce op is called, we can check to see if the queue has any results ready.

As part of the design, it would be important to make sure this behavior is optional, maybe by passing a flag to `DistributedOptimizer`.

Happy to chat more about this either on email or a Zoom call if you'd like.

Also, cc @romerojosh again, as we were just discussing this earlier tonight.

**Comment by Young768** [comment]
Hi @Tron-x my email is dyang33@uncc.edu
@tgaddair Yes, this design is optional. I think it has 3 cases. Lets say we have two processes p0 and p1. Normally, we should average gradients from these two processes, g0 and g1. If p0 is a slower one, then we might a zero tensor with g1. In the next iteration, we might average (g0' + g0) and g1 if p0 can catch up. Here, g0' is from the previous iteration.

**Comment by tgaddair** [comment]
Hey @Young768, to make sure I understand your proposal, if we suppose there are three processes {p0, p1, p2}, then we might imagine that at some time t0 the following gradients are ready to reduce:

p0: g0, g1
p1: g1
p2: g0

In this situation, are you suggesting that we allreduce g0 from p0 and p2, g1 from p0 and p1 without waiting for the other workers, then in the next step we incorporate the gradients from the other ranks?

**Comment by Young768** [comment]
@tgaddair Hi, I think what you suggest is similar to this paper from ASPLOS 2020 https://arxiv.org/abs/1909.08029. We can generate a new allreduce group for those ready ranks. Then in the next step, we incorporate other ranks. 

And what I suggest is similar to this paper from PPoPP 2020,  https://arxiv.org/pdf/1908.04207.pdf. We don't need to regenerate a new allreduce group. So if at step t0, only tensors from p0, p1 and p2 are ready, then other ranks participate the allreduce with a NULL tensor. From their results, I think both of these two work. Maybe we can start with a simple implementation.

**Comment by tgaddair** [comment]
Hey @Young768, thanks for sending along the papers.  I read through the second one (the solution you're proposing).  The idea sounds doable given Horovod's current approach.  I agree that most of the changes that need to be made will necessarily be in the Controller.  I would suggest taking a look at [this](https://github.com/horovod/horovod/pull/1058) PR, which added the Join operation to Horovod.  It similarly makes changes to the Controller and adds the ability for certain ranks to submit null / zero tensors to the allreduce.

There will likely be some additional gotchas to watch out for, as this would be effectively a kind of async SGD, but it would be interesting to see how it performs in situations with imbalanced workers.  Do you personally have a use case in mind where you think this could benefit?

I suspect in many cases, this approach will lead to convergence issues / less efficient fusion buffering, so we need make sure the code is not too tightly coupled with the "critical path" for the Horovod Controller.  We may be able to refactor parts of the Controller so that the eager vs traditional Controllers are separated from one another.

Also, given that the paper proposes both solo and majority eager SGD, is there one you are more interested in exploring?  I personally think the majority approach sounds the most useful at first glance, but curious to hear if you've tried it out and have any results to share.

**Comment by Young768** [comment]
@tgaddair Yep. I am looking into this PR recently. I did help me get familiar with the code. If I have further questions, I will email you.

Regarding the cases, in my mind, one is from imbalanced workloads across nodes. Another can be from the system heterogeneity.  For example, the cluster in my lab consists of machines with different capacities. Or if it is a shared cluster, it can be resource contention? How do you think?

Yep. I agree with you that this design is a tradeoff between statistical efficiency and system efficiency. I think it should be an optional configuration for users.

I am implementing the majority one. I will keep you updated.

**Comment by Young768** [comment]
Hi @tgaddair I just work back to this issue recently. I have two questions and maybe you can give me a hand.
1, I start with TF 2.0, when the cross-iteration is enabled, the slow worker will join the Allreduce with a null/zero tensor. What's your suggestion to implement suck a tensor? I can only find the zero tensor [here](https://www.tensorflow.org/api_docs/python/tf/zeros). And make sure the same shapes.

2, Could you please point out the code in Horovod that work as global barrier as mentioned above. In default, after each rank reports NEGOTIATE_ALLREDUCE, that rank should be locked and enters the next iteration after Allreduce. I am trying to look into the solution to remove the barrier.

Thanks

**Comment by Young768** [comment]
Hi @tgaddair @romerojosh I have a question on Tensorflow integrations to implement cross-iteration. Take the legacy optimizer as example:
`       

            gradients = self._optimizer.compute_gradients(*args, **kwargs)

            if size() > 1 or os.environ.get('HOROVOD_ELASTIC') == '1':
                grads, vars = zip(*gradients)
                avg_grads = self._allreduce_grads(grads)
                return list(zip(avg_grads, vars))
            else:
                return gradients
`

Suppose this process is the faster one. To avoid being blocked by waiting for Allreduce operation "avg_grads = self._allreduce_grads(grads)", can we just "return gradients" so it can continue the training with local gradients and enter the next step? But we need to develop new APIs to cache the gradients on the CPU and read the allreduced gradients from CPU if they have, instead of waiting for results from _allreduce_grads().
